File: frontend/src/api/gddContextStale.ts
/**
 * API — empreinte GDD et historique d'entité (Story 3.9 FR19).
 */
import apiClient from './client'

export interface GddContentFingerprintBody {
  context_selections: Record<string, unknown>
  field_configs?: Record<string, string[]>
  organization_mode?: string
}

export async function postGddContentFingerprint(
  body: GddContentFingerprintBody,
): Promise<{ fingerprint: string }> {
  const { data } = await apiClient.post<{ fingerprint: string }>(
    '/api/v1/context/gdd-content-fingerprint',
    body,
  )
  return data
}

export interface GddEntityHistoryEvent {
  at: string
  source: string
  summary: string
  snapshot?: Record<string, unknown> | null
  diff_from_previous?: string | null
}

export interface GddEntityHistoryResponse {
  category: string
  name: string
  events: GddEntityHistoryEvent[]
  diff_hint?: string | null
  previous_snapshot?: Record<string, unknown> | null
  current_snapshot?: Record<string, unknown> | null
}

export async function getGddEntityHistory(
  category: string,
  name: string,
  options?: { includeSnapshots?: boolean },
): Promise<GddEntityHistoryResponse> {
  const { data } = await apiClient.get<GddEntityHistoryResponse>(
    '/api/v1/context/gdd-entity-history',
    {
      params: {
        category,
        name,
        include_snapshots: options?.includeSnapshots === true,
      },
    },
  )
  return data
}

File: api/schemas/gdd_context_stale.py
"""Schémas API — empreinte GDD et historique d'entité (Story 3.9)."""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class GddContentFingerprintRequest(BaseModel):
    """Corps pour calculer l'empreinte du contenu GDD d'une sélection."""

    context_selections: Dict[str, Any] = Field(
        default_factory=dict,
        description="Sélections de contexte (même forme que generate-node / estimate-tokens)",
    )
    field_configs: Optional[Dict[str, List[str]]] = Field(
        None,
        description="Champs par type d'élément (optionnel, défauts ContextBuilder si absent)",
    )
    organization_mode: str = Field("narrative", description="Mode d'organisation du contexte")


class GddContentFingerprintResponse(BaseModel):
    """Réponse : empreinte SHA-256 du JSON de contexte structuré."""

    fingerprint: str = Field(..., description="Empreinte hex SHA-256")


class GddEntityHistoryEventPublic(BaseModel):
    """Événement d'historique ; snapshots optionnels selon la requête."""

    at: str
    source: str
    summary: str
    snapshot: Optional[Dict[str, Any]] = Field(
        None,
        description="Snapshot GDD à cette entrée si include_snapshots=true",
    )
    diff_from_previous: Optional[str] = Field(
        None,
        description="Diff texte vs l'entrée précédente (si include_snapshots et index > 0)",
    )


class GddEntityHistoryResponse(BaseModel):
    """Historique et diff MVP entre deux derniers snapshots."""

    category: str
    name: str
    events: List[GddEntityHistoryEventPublic]
    diff_hint: Optional[str] = Field(None, description="Texte de diff simplifié si ≥2 snapshots")
    previous_snapshot: Optional[Dict[str, Any]] = None
    current_snapshot: Optional[Dict[str, Any]] = None
